addon/query-cache.js

```diff
@@ -166,6 +171,11 @@ export default class QueryCache {
 
   _addResultToCache(result, cacheKey) {
     this._queryCache[cacheKey] = result;
+    //if the result is a promise
+    //skip adding to _reverseQueryCache
+    if (result instanceof Promise) {
```

Comment:
Instead of keeping two different types of data (the promise or the value) in the `_queryCache` structure, you can always keep the `Promise`. When the result has come back, the code can simply populate the reverse cache and leave the cache alone.

This should lead to simpler code, because it does not have to handle two different cases for `cachedValue`.

Reply:
Thank you @dnachev. It sounds good. I will take a look into this approach.

Reply:
I made necessary changes to clean up the logic to store the promise and update reverseCache upon success of ajax call.